### PR TITLE
gRPC support for TorchServe

### LINKED ISSUE #656: Add support for gRPC
Add support for gRPC in `TorchServe`

### PR DESCRIPTION
**Pull request by harshbafna**
## Description

 - Added gRPC support for inference and management APIs
   - inference.proto
   - management.proto
 - Added sample python grpc client in sanity suite and regression suite
 - Updated sanity suite to run APIs through both REST and gRPC APIs.
 - Added pytest cases in regresstion suite to validate all the inference and management APIs which were tested in newman suite, through gRPC apis
   - test_gRPC_inference_api.py
   - test_gRPC_management_apis.py

**gRPC vs REST performance report:** https://docs.google.com/spreadsheets/d/1Vo6ZmeMgWY1TPiNWPtM41tV8H7Vy8mUuBnG42hsPnbg/edit#gid=1514552587

Fixes #656

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

## Feature/Issue validation/testing

[gRPC_regression.log](https://github.com/pytorch/serve/files/5499422/gRPC_regression.log)

## Checklist:

- [x] Have you added tests that prove your fix is effective or that this feature works?
- [x] New and existing unit tests pass locally with these changes?
- [x] Has code been commented, particularly in hard-to-understand areas?
- [x] Have you made corresponding changes to the documentation?

### CODE REVIEW

**Comment by amtagrwl** [comment]
@harshbafna Is there an ETA on when this will be merged?

**Comment by harshbafna** [comment]
> @harshbafna Is there an ETA on when this will be merged?

@amtagrwl: The code is under review and will be merged once approved. :-)

File: docs/grpc_api.md
Comment (amtagrwl) [inline docs/grpc_api.md:7]: "../frontend/server/src/main/resources/proto/management.proto" should be changed to "../frontend/server/src/main/resources/proto/inference.proto"
```suggestion
* [Inference API](../frontend/server/src/main/resources/proto/inference.proto)
```

**Comment by maaquib** [comment]
@harshbafna ping for conflict resolution

**Comment by harshbafna** [comment]
> @harshbafna ping for conflict resolution

@maaquib : Done.

**Comment by chauhang** [comment]
@harshbafna This is a great start and will help customers to reduce e2e latency time for inferences. Please take care of following as well:

- Add integration tests in the postman tests, esp for the audio examples (waveglow/tactron) which will see th ebiggest benifit
- Add support for gRPC on secure channel. This can be done as a separate PR
- Update the results spreadsheet and add a column for the e2e time saved

**Comment by harshbafna** [comment]
@chauhang 

> * Add integration tests in the postman tests, esp for the audio examples (waveglow/tacotron) which will see the biggest benefit

We have added pytest cases using gRPC clients (we can not use postman/newman for gRPC requests). These test cases are data-driven and execute the same set of API calls which are used in postman tests for validating REST APIs.

[inference API test cases](https://github.com/pytorch/serve/pull/687/files#diff-37a6bc514a01c6998b924ee100836bdf56a711c49902fd391055406a4a6ccfd3R10)

[management API test cases](https://github.com/pytorch/serve/pull/687/files#diff-a2351339b680955c4fa272fdb7c107b9a8c6d3070e50d0c2396a754c75c030b9R9)

> * Add support for gRPC on secure channel. This can be done as a separate PR

We have added the hooks and flags for supporting gRPC on a secure channel. However, we were not able to connect to TorchServe using self-signed certificates and hence couldn't validate this. I will add a ticket to track this item.

> * Update the results spreadsheet and add a column for the e2e time saved

Updated.

File: frontend/server/src/main/resources/proto/inference.proto
Comment (harshbafna) [reply frontend/server/src/main/resources/proto/inference.proto:20]: @lxning: These are all HTTP layer params and gRPC is a wrapper on top of HTTP/2.0 protocol and all these parameters are populated by the gRPC framework.

Regarding API failures, we send back a proper gRPC error where all our existing HTTP error codes map to corresponding gRPC error codes and gRPC client gets an exception with the error message and status code.

Refer following code snippets

[Management api error response](https://github.com/pytorch/serve/pull/687/files/dc8e410e4b1ff891f361a970ea609947324ecc0d#diff-7e39f84a30c56df08b9bef1777ff8808f79978bb3999383bf5f945fa247a0474R152)

[Inference API error response](https://github.com/pytorch/serve/pull/687/files/dc8e410e4b1ff891f361a970ea609947324ecc0d#diff-cc7306545d600e99896faf2fd2d51075eb8154cc162ada158f8a9e85a2059d0dR110)

[HTTP to gRPC error code mapping](https://github.com/pytorch/serve/pull/687/files/dc8e410e4b1ff891f361a970ea609947324ecc0d#diff-0e64345b93bf7f4896722c6c29959cbf6a7bee6fcc15e9cf3f0ad68f81d463f5R23)

File: frontend/server/src/main/resources/proto/inference.proto
Comment (harshbafna) [reply frontend/server/src/main/resources/proto/inference.proto:9]: Please refer my inputs on the other [comment](https://github.com/pytorch/serve/pull/687/files/dc8e410e4b1ff891f361a970ea609947324ecc0d#r516428720)

**Comment by chauhang** [comment]
@harshbafna The regressions tests are failing for gRPC brach. Please see the logs from @HamidShojanazeri  tests:

[regressin_grpc.logs.txt](https://github.com/pytorch/serve/files/5488067/regressin_grpc.logs.txt)

**Review by chauhang — CHANGES_REQUESTED** [review]
Please resolve the regression test issues. Also link up the new GH issue that you have created for the Secure channel support problems

**Comment by harshbafna** [comment]
> Please resolve the regression test issues. Also link up the new GH issue that you have created for the Secure channel support problems

@chauhang 

Created following ticket for gRPC Secure Channel support:
https://github.com/pytorch/serve/issues/776

The regression test suite is failing with `newman` command not found, which is a known issue in master as well and PR is already in place for this: https://github.com/pytorch/serve/pull/757

**Comment by harshbafna** [comment]
> Please resolve the regression test issues. Also link up the new GH issue that you have created for the Secure channel support problems

Regression test issues resolved. Attached latest logs in the PR description.

**Comment by dhanainme** [comment]
@harshbafna  / @dhaniram-kshirsagar can you please fix the merge conflicts in this PR.

**Review by chauhang — APPROVED** [review]

**Comment by dhanainme** [comment]
@harshbafna Can you please take a look @ the merge conflicts

File: frontend/server/src/main/java/org/pytorch/serve/ModelServer.java
Comment (lxning) [inline frontend/server/src/main/java/org/pytorch/serve/ModelServer.java:114]: it is better to be configurable: REST/gRPC/both

Comment (harshbafna) [reply frontend/server/src/main/java/org/pytorch/serve/ModelServer.java:114]: @lxning: Logged following enhancement ticket to track this: #799. This is not a blocker item for the 0.3.0 release.

**Review by lxning — APPROVED** [review]

File: frontend/server/src/main/java/org/pytorch/serve/wlm/WorkerThread.java
Comment (maaquib) [inline frontend/server/src/main/java/org/pytorch/serve/wlm/WorkerThread.java:450]: Same as above, can you explain whats the reason for change 507 to 513?

**Review by maaquib — CHANGES_REQUESTED** [review]
Review comments inline

File: README.md
Comment (harshbafna) [reply README.md:146]: Already a part of developer.txt file. However, other packages are not required in case a user is just planning to implement a TorchServe gRPC client.

File: frontend/server/src/main/java/org/pytorch/serve/grpcimpl/ManagementImpl.java
Comment (harshbafna) [reply frontend/server/src/main/java/org/pytorch/serve/grpcimpl/ManagementImpl.java:163]: @maaquib: Both Inference APIs and Management APIs use different responseObserver types. Also, the inference API has only one main API (predict). Thus for avoiding one minor code duplication we will need to add a bunch of complicated type castings.

I have refactored the InferenceImpl.java to avoid code duplication within the file for sending error response.

File: frontend/server/src/main/java/org/pytorch/serve/job/RestJob.java
Comment (harshbafna) [reply frontend/server/src/main/java/org/pytorch/serve/job/RestJob.java:48]: This was changed as part of #666. At the time of job queue both `begin` and `scheduled` hold the same value. Also, the `scheduled` parameter provides a setter that will overwrite the value with `System.nanoTime();` whenever called.

File: frontend/server/src/main/java/org/pytorch/serve/job/RestJob.java
Comment (harshbafna) [reply frontend/server/src/main/java/org/pytorch/serve/job/RestJob.java:103]: This was because we refactored all the error codes sent from the `WLM` package to use HTTP Status code  Java's inbuilt `HTTPURLConnectio`n instead of Netty's `HTTPResponseStatus`. And `HTTPURLConnection` doesn't has the `507` code available. Hence, we used the `413` code instead and re-mapped it to `507` to keep the error codes backward compatible.

File: frontend/server/src/main/java/org/pytorch/serve/wlm/WorkerThread.java
Comment (harshbafna) [reply frontend/server/src/main/java/org/pytorch/serve/wlm/WorkerThread.java:450]: HttpURLConnection doesn't support 507 code.

File: frontend/server/src/main/java/org/pytorch/serve/grpcimpl/InferenceImpl.java
Comment (harshbafna) [reply frontend/server/src/main/java/org/pytorch/serve/grpcimpl/InferenceImpl.java:84]: Done.

File: frontend/server/src/main/java/org/pytorch/serve/grpcimpl/InferenceImpl.java
Comment (harshbafna) [reply frontend/server/src/main/java/org/pytorch/serve/grpcimpl/InferenceImpl.java]: Done.

File: frontend/server/src/main/java/org/pytorch/serve/grpcimpl/InferenceImpl.java
Comment (harshbafna) [reply frontend/server/src/main/java/org/pytorch/serve/grpcimpl/InferenceImpl.java]: Done.

File: frontend/build.gradle
Comment (harshbafna) [reply frontend/build.gradle]: Done.

File: docs/inference_api.md
Comment (harshbafna) [reply docs/inference_api.md]: Removed.

File: docs/configuration.md
Comment (harshbafna) [reply docs/configuration.md]: Done.

**Comment by maaquib** [comment]
@harshbafna Please resolve the conflicts on this. Thanks

**Comment by harshbafna** [comment]
> @harshbafna Please resolve the conflicts on this. Thanks

@maaquib Done.

**Comment by maaquib** [comment]
Failing regression tests
```
## DLAMI 38
git clone https://github.com/pytorch/serve/
cd serve/
git checkout issue_656
python ts_scripts/install_dependencies.py --environment dev
python torchserve_sanity.py
python test/regression_tests.py
```

```
## Successfully started TorchServe
newman

management_api_collection

→ management request
  {{METHOD}} http://localhost:8081/{{path}} [errored]
     Method must be a valid HTTP token ["{{METHOD}}"]
  2. Successful  request

┌─────────────────────────┬──────────┬──────────┐
│                         │ executed │   failed │
├─────────────────────────┼──────────┼──────────┤
│              iterations │        1 │        0 │
├─────────────────────────┼──────────┼──────────┤
│                requests │        1 │        1 │
├─────────────────────────┼──────────┼──────────┤
│            test-scripts │        1 │        0 │
├─────────────────────────┼──────────┼──────────┤
│      prerequest-scripts │        0 │        0 │
├─────────────────────────┼──────────┼──────────┤
│              assertions │        1 │        1 │
├─────────────────────────┴──────────┴──────────┤
│ total run duration: 75ms                      │
├───────────────────────────────────────────────┤
│ total data received: 0B (approx)              │
└───────────────────────────────────────────────┘

  #  failure                          detail

 1.  TypeError                        Method must be a valid HTTP token ["{{METHOD}}"]
                                      at request
                                      inside "management request"

 2.  AssertionError                   Successful  request
                                      expected undefined to be one of [ 200 ]
                                      at assertion:0 in test-script
                                      inside "management request"
## Stopping TorchServe
## In directory: /home/ubuntu/serve/test | Executing command: torchserve --stop
TorchServe has stopped.
```

**Review by maaquib — CHANGES_REQUESTED** [review]
@harshbafna Failing regression tests
```
## DLAMI 38, g4dn.xl
git clone https://github.com/pytorch/serve/
cd serve/
git checkout issue_656
python ts_scripts/install_dependencies.py --environment dev
python torchserve_sanity.py
python test/regression_tests.py
```

```
## Successfully started TorchServe
newman

management_api_collection

→ management request
  {{METHOD}} http://localhost:8081/{{path}} [errored]
     Method must be a valid HTTP token ["{{METHOD}}"]
  2. Successful  request

┌─────────────────────────┬──────────┬──────────┐
│                         │ executed │   failed │
├─────────────────────────┼──────────┼──────────┤
│              iterations │        1 │        0 │
├─────────────────────────┼──────────┼──────────┤
│                requests │        1 │        1 │
├─────────────────────────┼──────────┼──────────┤
│            test-scripts │        1 │        0 │
├─────────────────────────┼──────────┼──────────┤
│      prerequest-scripts │        0 │        0 │
├─────────────────────────┼──────────┼──────────┤
│              assertions │        1 │        1 │
├─────────────────────────┴──────────┴──────────┤
│ total run duration: 75ms                      │
├───────────────────────────────────────────────┤
│ total data received: 0B (approx)              │
└───────────────────────────────────────────────┘

  #  failure                          detail

 1.  TypeError                        Method must be a valid HTTP token ["{{METHOD}}"]
                                      at request
                                      inside "management request"

 2.  AssertionError                   Successful  request
                                      expected undefined to be one of [ 200 ]
                                      at assertion:0 in test-script
                                      inside "management request"
## Stopping TorchServe
## In directory: /home/ubuntu/serve/test | Executing command: torchserve --stop
TorchServe has stopped.
```

**Comment by harshbafna** [comment]
> @harshbafna Failing regression tests
> 
> ```
> ## DLAMI 38, g4dn.xl
> git clone https://github.com/pytorch/serve/
> cd serve/
> git checkout issue_656
> python ts_scripts/install_dependencies.py --environment dev
> python torchserve_sanity.py
> python test/regression_tests.py
> ```
> 
> ```
> ## Successfully started TorchServe
> newman
> 
> management_api_collection
> 
> → management request
>   {{METHOD}} http://localhost:8081/{{path}} [errored]
>      Method must be a valid HTTP token ["{{METHOD}}"]
>   2. Successful  request
> 
> ┌─────────────────────────┬──────────┬──────────┐
> │                         │ executed │   failed │
> ├─────────────────────────┼──────────┼──────────┤
> │              iterations │        1 │        0 │
> ├─────────────────────────┼──────────┼──────────┤
> │                requests │        1 │        1 │
> ├─────────────────────────┼──────────┼──────────┤
> │            test-scripts │        1 │        0 │
> ├─────────────────────────┼──────────┼──────────┤
> │      prerequest-scripts │        0 │        0 │
> ├─────────────────────────┼──────────┼──────────┤
> │              assertions │        1 │        1 │
> ├─────────────────────────┴──────────┴──────────┤
> │ total run duration: 75ms                      │
> ├───────────────────────────────────────────────┤
> │ total data received: 0B (approx)              │
> └───────────────────────────────────────────────┘
> 
>   #  failure                          detail
> 
>  1.  TypeError                        Method must be a valid HTTP token ["{{METHOD}}"]
>                                       at request
>                                       inside "management request"
> 
>  2.  AssertionError                   Successful  request
>                                       expected undefined to be one of [ 200 ]
>                                       at assertion:0 in test-script
>                                       inside "management request"
> ## Stopping TorchServe
> ## In directory: /home/ubuntu/serve/test | Executing command: torchserve --stop
> TorchServe has stopped.
> ```

@maaquib Fixed the issue. 
Logs: [gRPC regression.txt](https://github.com/pytorch/serve/files/5670595/gRPC.regression.txt)

**Comment by dhanainme** [comment]
Can confirm that the regression tests are passing for both CPU & GPU in CodeBuild env.

Screen Shot 2020-12-10 at 5 04 23 AM
Screen Shot 2020-12-10 at 5 03 58 AM

**Review by dhanainme — APPROVED** [review]

**Comment by dhanainme** [comment]
@maaquib i can go ahead and merge this if it looks good to you.

**Review by maaquib — APPROVED** [review]

**Pull request merged**
